tslint/relayOperationGenericsRule.js

```diff
@@ -0,0 +1,277 @@
+            typeArgument,
+            `expected operation type parameter to be \`${operationName}\``,
+            fixes
+          )
+        }
+      }
+    } else {
+      this.addFailureAtNode(
+        taggedTemplate,
+        "expected a graphql`…` tagged-template"
+      )
+    }
+  }
+
+  /**
+   * @param {number} start
+   * @param {number} width
+   * @param {string} replacement
+   * @param {string} operationName
+   */
+  createFixes(start, width, replacement, operationName) {
+    const fixes = [new Lint.Replacement(start, width, replacement)]
+    if (!this.hasImportForOperation(operationName)) {
+      fixes.push(this.importDeclarationFixForOperation(operationName))
+    }
+    return new Lint.Fix("fix-relay-type-parameter", fixes)
+  }
+
+  /**
+   * @param {string} operationName
+   */
+  importPathForOperation(operationName) {
+    const options = this.getOptions()[0] || {
+      artifactDirectory: "__generated__",
+      makeRelative: false,
+    }
+    if (options.makeRelative) {
+      throw new Error(
+        "[relayOperationGenericsRule] Making import declarations relative is not implemented yet."
+      )
+    }
+    return `${options.artifactDirectory}/${operationName}.graphql`
+  }
+
+  importDeclarationFixForOperation(operationName) {
+    const path = this.importPathForOperation(operationName)
+    const importDeclaration = `import { ${operationName} } from "${path}"\n`
+
+    const imports = this.getImports()
+    const lastImport = imports[imports.length - 1]
+
+    let start = 0
+    if (lastImport) {
+      start = lastImport.getEnd() + 1
+    }
+
+    return new Lint.Replacement(start, 0, importDeclaration)
+  }
+
+  /**
+   * @param {string} operationName
+   */
+  hasImportForOperation(operationName) {
+    // TODO: So many hoops to jump through without TS :/
+    /** @type {any} */
+    let any
+
+    const importPath = this.importPathForOperation(operationName)
+
+    return this.getImports().some(node => {
+      any = node.moduleSpecifier
+      /** @type {ts.StringLiteral} */
+      const path = any
+      if (path.text === importPath && node.importClause) {
+        any = node.importClause.namedBindings
+        /** @type {ts.NamedImports} */
+        const namedBindings = any
+        if (namedBindings) {
+          return namedBindings.elements.some(
+            element => element.name.getText() === operationName
+          )
+        }
+      }
+      return false
+    })
+  }
+}
+
+/**
+ *
+ * @param {ts.TaggedTemplateExpression} taggedTemplate
+ * @return {string | null}
+ */
+function getOperationName(taggedTemplate) {
+  const template = taggedTemplate.template.getFullText()
+  // Strip backticks
+  const source = template.substring(1, template.length - 1)
+
+  const ast = parse(source)
+  let queryName = null
+  visit(ast, {
+    OperationDefinition(node) {
+      queryName = node.name.value
+      return BREAK
+    },
+  })
+
+  return queryName
+}
+
+module.exports = { Rule }
```

Comment:
💯Very nice -- would be cool to do a L&L on this to cover some of the basics of an AST and walking it (thinking something in general since the principles also apply to things like Babel and co).

Reply:
Great idea! 👍